File: quests/quest00/ex04.js
"use strict";
// `cat my_z` must print Z followed by a newline. 2 points.

const FILE = "my_z";
const WANTED = "Z\n";

exports.run = function (g) {
  const exists = g.check(
    `${FILE} exists`,
    g.exists(FILE),
    g.exists(FILE) ? "" : "no such file in this directory"
  );

  if (!exists) {
    g.check(`cat ${FILE} prints Z`, false, "nothing to cat");
    return;
  }

  const { code, stdout, stderr } = g.sh(`cat ${FILE}`);
  const out = stdout.toString();
  const ok = code === 0 && out === WANTED;
  const detail =
    code !== 0
      ? `cat exited with ${code}: ${stderr.toString().trim()}`
      : ok
      ? ""
      : `got ${JSON.stringify(out)}, expected ${JSON.stringify(WANTED)}`;
  g.check(`cat ${FILE} prints Z`, ok, detail);
};

File: quests/quest00/ex04.py
"""ex04 - `cat my_z` must print Z followed by a newline. 2 points."""

FILE = "my_z"
WANTED = b"Z\n"


def run(g):
    exists = g.check("%s exists" % FILE,
                     g.exists(FILE),
                     "" if g.exists(FILE) else "no such file in this directory")

    if not exists:
        g.check("cat %s prints Z" % FILE, False, "nothing to cat")
        return

    code, out, err = g.sh("cat %s" % FILE)
    ok = code == 0 and out == WANTED
    if code != 0:
        detail = "cat exited with %d: %s" % (code, err.decode("utf-8", "replace").strip())
    else:
        detail = "" if ok else "got %r, expected %r" % (out, WANTED)
    g.check("cat %s prints Z" % FILE, ok, detail)
